src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/ErrorListInProcess.cs

```diff
@@ -50,53 +49,81 @@ public async Task<ImmutableArray<string>> GetErrorsAsync(ErrorSource errorSource
+            var list = errorItems.Select(GetMessage).ToList();
 
             return list
                 .OrderBy(x => x, StringComparer.OrdinalIgnoreCase)
                 .ThenBy(x => x, StringComparer.Ordinal)
                 .ToImmutableArray();
         }
 
-        private async Task<ImmutableArray<ITableEntryHandle>> GetErrorItemsAsync(CancellationToken cancellationToken)
+        public async Task<string> NavigateToErrorListItemAsync(int item, bool isPreview, bool shouldActivate, CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var errorList = await GetRequiredGlobalServiceAsync<SVsErrorList, IErrorList>(cancellationToken);
+            ErrorSource errorSource = 0;
+            if (errorList.AreBuildErrorSourceEntriesShown)
+                errorSource |= ErrorSource.Build;
+
+            if (errorList.AreOtherErrorSourceEntriesShown)
+                errorSource |= ErrorSource.Other;
+
+            var minimumSeverity =
+                errorList.AreMessagesShown ? __VSERRORCATEGORY.EC_MESSAGE :
+                errorList.AreWarningsShown ? __VSERRORCATEGORY.EC_WARNING :
+                __VSERRORCATEGORY.EC_ERROR;
+
+            var items = await GetErrorItemsAsync(errorSource, minimumSeverity, cancellationToken);
+
+            items[item].NavigateTo(isPreview, shouldActivate);
+            return GetMessage(items[item]);
+        }
+
+        private async Task<ImmutableArray<ITableEntryHandle>> GetErrorItemsAsync(ErrorSource errorSource, __VSERRORCATEGORY minimumSeverity, CancellationToken cancellationToken)
         {
             await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
 
             var errorList = await GetRequiredGlobalServiceAsync<SVsErrorList, IErrorList>(cancellationToken);
             var args = await errorList.TableControl.ForceUpdateAsync().WithCancellation(cancellationToken);
-            return args.AllEntries.ToImmutableArray();
+            return args.AllEntries
+                .Where(item =>
+                {
+                    if (item.GetCategory() > minimumSeverity)
+                    {
+                        return false;
+                    }
+
+                    if (item.GetErrorSource() is not { } itemErrorSource
+                        || !errorSource.HasFlag(itemErrorSource))
+                    {
+                        return false;
+                    }
+
+                    return true;
+                })
+                .ToImmutableArray();
+        }
+
+        private static string GetMessage(ITableEntryHandle item)
+        {
+            var source = item.GetBuildTool();
+            var document = Path.GetFileName(item.GetPath() ?? item.GetDocumentName()) ?? "<unknown>";
+            var line = item.GetLine() ?? -1;
+            var column = item.GetColumn() ?? -1;
```

Comment:
This is moved from a previous location that did the same thing. So far it hasn't been a problem.

There are a few ways it could be changed to be less code, but none of them seem to be easier to understand.